### PR TITLE
fix(search): Make --all work for catalog

### PR DESCRIPTION
**Pull request by dcarley**
## Proposed Changes

**fix(search): Use pagination**

So that `--all` works. This however breaks the behaviour when `--all` is
not specified, by returning all results, which will be fixed in a
subsequent commit.

**fix(search): Add limit**

So that we only return up to the limit of results when `--all` is not
specified. This however breaks the hint that more results are available,
which will be fixed in a subsequent commit.

**fix(search): Return correct count**

This fixes the hint that more results are available when `--all` is not
specified and the results exceed the limit.

To do this we needed access to `total_count` from one of the responses
which didn't fit terribly nicely with the streaming results.

## Release Notes

N/A until catalog is released.

### CODE REVIEW

File: cli/flox-rust-sdk/src/providers/catalog.rs [hunk 1]
```diff
@@ -286,40 +288,66 @@ impl ClientTrait for CatalogClient {
...
+                Ok::<_, SearchError>((
+                    packages.total_count,
+                    packages
+                        .items
+                        .into_iter()
+                        .map(TryInto::<SearchResult>::try_into)
+                        .collect::<Result<Vec<_>, _>>()?,
+                ))
+            },
+            PAGE_SIZE,
```
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/catalog.rs:327]: 🚧 Increasing the default page size to something like 100 could improve the experience for large searches but we'd want to take the lower of `limit || PAGE_SIZE` to ensure that we don't fetch too many results.

File: cli/flox-rust-sdk/src/providers/catalog.rs [hunk 2]
```diff
@@ -243,6 +243,8 @@ pub trait ClientTrait {
     ) -> Result<SearchResults, VersionsError>;
 }
 
+const PAGE_SIZE: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(10) };
+
 #[async_trait]
 impl ClientTrait for CatalogClient {
```
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/catalog.rs:249]: 🚧 There isn't any unit or integration test coverage for the real client methods in here. I see that progenitor can [generate httpmock helpers](https://docs.rs/progenitor/latest/progenitor/struct.Generator.html#method.httpmock), which might be useful for adding coverage in the future.

File: cli/flox-rust-sdk/src/providers/catalog.rs [hunk 3]
```diff
@@ -243,6 +243,8 @@ pub trait ClientTrait {
     ) -> Result<SearchResults, VersionsError>;
 }
 
+const PAGE_SIZE: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(10) };
```
Comment (mkenigs) [inline cli/flox-rust-sdk/src/providers/catalog.rs:246]: Did you experiment at all with different page sizes?

File: cli/flox/src/commands/search.rs [hunk 4]
```diff
@@ -74,15 +74,8 @@ impl Search {
...
             client
-                .search(
-                    &self.search_term,
-                    flox.system.clone(),
-                    limit.unwrap_or(
-                        DEFAULT_SEARCH_LIMIT.expect("default search limit didn't have a value"),
-                    ),
-                )
+                .search(&self.search_term, flox.system.clone(), limit)
                 .await?
```
Comment (dcarley) [inline cli/flox/src/commands/search.rs:79]: 🚧 Wrapping this in a spinner would improve the user experience for slow searches that take many pages but I couldn't see any way to do this for something that returns a future. Maybe requires a new ticket to extend the spinner to support futures?

File: cli/flox-rust-sdk/src/providers/catalog.rs [hunk 5]
```diff
@@ -753,9 +813,14 @@ mod tests {
...
-        let collected: Vec<i32> = stream.try_collect().await.unwrap();
+        let collected: Vec<StreamItem<i32>> = stream.try_collect().await.unwrap();
 
-        assert_eq!(collected, (1..=3).collect::<Vec<_>>());
+        assert_eq!(collected, [
+            StreamItem::TotalCount(3),
+            StreamItem::Result(1),
+            StreamItem::Result(2),
+            StreamItem::Result(3)
+        ]);
```
Comment (dcarley) [inline cli/flox-rust-sdk/src/providers/catalog.rs:823]: 🚧 I gave up finding the funcitonal way to do this and the other tests need updating to include TotalCount.

File: cli/flox/src/commands/search.rs
Comment (dcarley) [inline cli/flox/src/commands/search.rs:1]: Example of this working:

```
~/projects/flox/flox/cli on dcarley/1457-catalog-search-pagination [$!><]
% FLOX_FEATURES_USE_CATALOG=true just flox search nodejs 
make: Entering directory '/Users/dcarley/projects/flox/flox/pkgdb'
mkdir -p tests;
Generating tests/.gitignore
make: Leaving directory '/Users/dcarley/projects/flox/flox/pkgdb'
~/projects/flox/flox/cli ~/projects/flox/flox
nodejs       Event-driven I/O framework for the V8 JavaScript engine
nodejs_22    Event-driven I/O framework for the V8 JavaScript engine
nodejs_21    Event-driven I/O framework for the V8 JavaScript engine
nodejs_20    Event-driven I/O framework for the V8 JavaScript engine
nodejs_18    Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim  Event-driven I/O framework for the V8 JavaScript engine
nodejs-19_x  Event-driven I/O framework for the V8 JavaScript engine
nodejs-18_x  Event-driven I/O framework for the V8 JavaScript engine
nodejs-16_x  Event-driven I/O framework for the V8 JavaScript engine
nodejs-14_x  Event-driven I/O framework for the V8 JavaScript engine

Showing 10 of 47 results. Use `flox search nodejs --all` to see the full list.

Use 'flox show <package>' to see available versions


~/projects/flox/flox/cli on dcarley/1457-catalog-search-pagination [$!><] took 10s
% FLOX_FEATURES_USE_CATALOG=true just flox search nodejs --all
make: Entering directory '/Users/dcarley/projects/flox/flox/pkgdb'
mkdir -p tests;
Generating tests/.gitignore
make: Leaving directory '/Users/dcarley/projects/flox/flox/pkgdb'
~/projects/flox/flox/cli ~/projects/flox/flox
nodejs                                           Event-driven I/O framework for the V8 JavaScript engine
nodejs_22                                        Event-driven I/O framework for the V8 JavaScript engine
nodejs_21                                        Event-driven I/O framework for the V8 JavaScript engine
nodejs_20                                        Event-driven I/O framework for the V8 JavaScript engine
nodejs_18                                        Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim                                      Event-driven I/O framework for the V8 JavaScript engine
nodejs-19_x                                      Event-driven I/O framework for the V8 JavaScript engine
nodejs-18_x                                      Event-driven I/O framework for the V8 JavaScript engine
nodejs-16_x                                      Event-driven I/O framework for the V8 JavaScript engine
nodejs-14_x                                      Event-driven I/O framework for the V8 JavaScript engine
nodejs_latest                                    Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim_22                                   Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim_21                                   Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim_20                                   Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim_18                                   Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim-19_x                                 Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim-18_x                                 Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim-16_x                                 Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim-14_x                                 Event-driven I/O framework for the V8 JavaScript engine
nodejs-slim_latest                               Event-driven I/O framework for the V8 JavaScript engine
elmPackages.nodejs                               Event-driven I/O framework for the V8 JavaScript engine
nodePackages.nodejs                              Event-driven I/O framework for the V8 JavaScript engine
nodejs-16_x-openssl_1_1                          Event-driven I/O framework for the V8 JavaScript engine
haxePackages.hxnodejs_6                          Extern definitions for node.js 6.9
haxePackages.hxnodejs_4                          Extern definitions for node.js 4.x
matrix-sdk-crypto-nodejs                         A no-network-IO implementation of a state machine that handles E2EE for Matrix clients
graalvmCEPackages.graalnodejs                    High-Performance Polyglot VM (Product: nodejs-installable-svm)
dockerfile-language-server-nodejs                A language server for Dockerfiles powered by Node.js, TypeScript, and VSCode technologies
python39Packages.hatch-nodejs-version            Plugins for dealing with NodeJS versions
matrix-sdk-crypto-nodejs-0_1_0-beta_3            A no-network-IO implementation of a state machine that handles E2EE for Matrix clients
python312Packages.hatch-nodejs-version           Plugins for dealing with NodeJS versions
python311Packages.hatch-nodejs-version           Plugins for dealing with NodeJS versions
python310Packages.hatch-nodejs-version           Plugins for dealing with NodeJS versions
nodePackages.dockerfile-language-server-nodejs   A language server for Dockerfiles powered by Node.js, TypeScript, and VSCode technologies
graalvmCEPackages.nodejs-installable-svm-java19  High-Performance Polyglot VM (Product: nodejs-installable-svm)
graalvmCEPackages.nodejs-installable-svm-java17  High-Performance Polyglot VM (Product: nodejs-installable-svm)
graalvmCEPackages.nodejs-installable-svm-java11  High-Performance Polyglot VM (Product: nodejs-installable-svm)
corepack                                                 Wrappers for npm, pnpm and Yarn via Node.js Corepack
corepack_22                                              Wrappers for npm, pnpm and Yarn via Node.js Corepack
corepack_21                                              Wrappers for npm, pnpm and Yarn via Node.js Corepack
corepack_20                                              Wrappers for npm, pnpm and Yarn via Node.js Corepack
corepack_18                                              Wrappers for npm, pnpm and Yarn via Node.js Corepack
corepack_latest                                          Wrappers for npm, pnpm and Yarn via Node.js Corepack
nodenv                                                   Manage multiple NodeJS versions
nodehun                                                  The Hunspell binding for NodeJS that exposes as much of Hunspell as possible and also adds new features
nodePackages.svgo                                        Nodejs-based tool for optimizing SVG vector graphics files
nodePackages.@gitbeaker/cli                              Full NodeJS CLI implementation of the GitLab API.

Use 'flox show <package>' to see available versions


~/projects/flox/flox/cli on dcarley/1457-catalog-search-pagination [$!><] took 17s
%
```

**Review by dcarley — COMMENTED** [review]
@mkenigs @zmitchell can you see this through to completion for me while I'm on hoilday? 🙏🏻

**Comment by mkenigs** [comment]
Closing in favor of https://github.com/flox/flox/pull/1498

**Pull request closed without merging**